bootstrap/bootstrap.go

```diff
@@ -173,8 +195,111 @@ func (b *Bootstrap) Cancel() error {
 	return nil
 }
 
+// extractTraceCtx pulls encoded distributed tracing information from the env vars.
+// Note: This should match the injectTraceCtx code in shell.
+func (b *Bootstrap) extractTraceCtx() opentracing.SpanContext {
+	// TODO: fallback to BUILD_ID/JOB_ID
+	traceCtx, ok := b.shell.Env.Get("BUILDKITE_TRACE_CONTEXT")
+	if !ok {
+		// Return nil so a new span will be created
+		return nil
+	}
+	decodedCtx, err := base64.URLEncoding.DecodeString(traceCtx)
+	if err != nil {
+		b.shell.Warningf("trace extraction decode error: %v", err)
+		// Not worth stopping the run over
+		return nil
+	}
+	dec := gob.NewDecoder(bytes.NewBuffer(decodedCtx))
+	carrier := tracer.TextMapCarrier{}
+	if err := dec.Decode(&carrier); err != nil {
+		b.shell.Warningf("tracing carrier extraction decoding error: %v", err)
+		// Not worth stopping the run over
+		return nil
+	}
+
+	wireContext, err := opentracing.GlobalTracer().Extract(
+		opentracing.TextMap,
+		carrier,
+	)
+	if err != nil {
+		b.shell.Warningf("trace extraction error: %v", err)
+		// Not worth stopping the run over
+		return nil
+	}
+
+	return wireContext
+}
+
+// stopper lets us abstract the tracer wrap up code so we can plug in different tracing
+// library implementations that are opentracing compatible. Opentracing itself
+// doesn't have a Stop function on its Tracer interface.
+type stopper func()
+
+// startTracing sets up tracing based on the config values. It uses opentracing as an
+// abstraction so the agent can support multiple libraries if needbe.
+func (b *Bootstrap) startTracing(ctx context.Context) (opentracing.Span, context.Context, stopper) {
+	b.shell.Commentf("Setting up tracer.")
+	buildID, _ := b.shell.Env.Get("BUILDKITE_BUILD_ID")
+	source, _ := b.shell.Env.Get("BUILDKITE_SOURCE")
+	label, hasLabel := b.shell.Env.Get("BUILDKITE_LABEL")
+	if !hasLabel {
+		label = "job"
+	}
+
+	// Set specific tracing library here. Everything else should be using opentracing.
+	// Use a constant sampler - CI runs aren't high traffic.
+	var t opentracing.Tracer
+	var stopper stopper
+	if b.Config.TracingDatadogAddr != "" {
+		t = opentracer.New(
+			tracer.WithAgentAddr("localhost:8126"),
+			tracer.WithServiceName("buildkite_agent"),
+			tracer.WithSampler(tracer.NewAllSampler()),
+			tracer.WithGlobalTag("buildkite.agent", b.AgentName),
+			tracer.WithGlobalTag("buildkite.queue", b.Queue),
+			tracer.WithGlobalTag("buildkite.pipeline", b.PipelineSlug),
+			tracer.WithGlobalTag("buildkite.branch", b.Branch),
+			tracer.WithGlobalTag("buildkite.job_id", b.JobID),
+			tracer.WithGlobalTag("buildkite.build_id", buildID),
+			tracer.WithGlobalTag("buildkite.source", source),
```

Comment:
Maybe some of these shouldn't be _global_ tags (they'll appear in every span)?

Reply:
Seems useful to have this information on all spans, but I don't have a deep familiarity with app analytics and how the search functionality works. Maybe @samschlegel can weigh in on best practices on this one.